### ISSUE TITLE
Implement new filter action "Forward with SmartTemplate"

### ISSUE THREAD

**Issue by RealRaven2000** [RealRaven2000/FiltaQuilla#153]
This would be very useful for mass-forwarding / processing of Emails. A new action

"Forward with SmartTemplates"

and also

"Reply with SmartTemplates"

In order to minimize work on the side of SmartTemplates this would have to use the current workflow (open composer while triggering SmartTemplates to do its work, such as subject modification, adding recipients, text replacements and cleaning of quoted material) and automate it - it would have to send off the email after SmartTemplates has finished its processing.

Related SmartTemplates issue: [issue 173](https://github.com/RealRaven2000/SmartTemplates/issues/173)

**Comment by RafalGemex** [comment]
The Issue: When email gets forwarded the details of the original senders are being added to the header of the forwarded message

USER1 and USER2 are human persons sending emails between each other
BRIDGE is a designated computer with Thunderbird installed and Message Filters applied
 Message filters in place:
       1. If sender is USER1 forward the message to USER2
       2. If sender is USER2 forward the message to USER1

Regarding the attached:
1. USER2 sends an email to the BRIDGE
2. BRIDGE is set up to automatically forward the message, the forwarded message contains details of the sender

Solution:  The message gets forwarded automatically while removing the header with the details of the original sender

Extra info: We have a designated computer with designated email address so any flashing up composer windows will not be a problem as the computer is a stand alone machine set up to only forward the emails


![Sample](https://user-images.githubusercontent.com/100281749/155307585-7e3c9323-14b7-4c30-bfaa-97c6c832cdfa.png)

**Comment by RealRaven2000** [comment]
> Regarding the attached:
> 
> 1. USER2 sends an email to the BRIDGE
> 2. BRIDGE is set up to automatically forward the message, the forwarded message **_contains details of the sender_**

if the body contains the email address / name you can use the function `%header.set.matchFromBody()%`

for example, if you have a line followed by name and email, such as:

Customer
------------------------------------------------------------------
Fred Flintsone
Acme Stonebreakers INc.
fred.flint@stonebreakers.com

you can this to extract and set the new "to" address:

`%header.set.matchFromBody(to,"Customer.*\n\-+\n.*\n(.*\n){0,2}(.*\@.*)\n",2)%`

For addressing USER1 from the same example you could use:

```
<p>Dear %matchTextFromBody("Customer.*\n\-+\n(.*)\n",1)%,</p>

```

> Solution: The message gets forwarded automatically while removing the header with the details of the original sender
you use the word "header" - I think what you are meaning is the "quote header", which usually reads something like this:

<html><body>-------- Forwarded Message --------

Subject: | multiple mail test (8)
-- | --
Fri, 28 Jan 2022 13:51:55 +0000
Thunderbird Daily-Addons <thunderbirddaily@gmail.com>
Thunderbird Daily-Addons <thunderbirddaily@gmail.com>



As standard this is replaced by anything you like (configured through account settings in SmartTemplates)

![image](https://user-images.githubusercontent.com/6314674/155312050-4b531b81-c416-485c-b85a-47eeeab3e2e4.png)

please let's call this "quote header" for future to avoid confusion. We use the term _header_ for any field in the email (such as subject line, to, from, date sent etc.)

**Comment by RealRaven2000** [comment]
> 1. If sender is USER1 forward the message to USER2
> 2. If sender is USER2 forward the message to USER1

One more question, is it just the 2 senders and you have a fixed one-on-one relationship? This way the filter would know about who it's from and who to send it to and this would be hard coded in the filter actions. However - you cannot set the recipient as part of the filter condition (forward using - "template XX") because you cannot add another parameter (to USER2). If you cannot extract the "USER2" information from the actual email (as I demonstrated in the example above) then you will need a dedicated template that has hardcoded the to address such as: `%header.set(to,"USER2 <user2@zzmail.com")%`  - that would also mean you need a dedicated template file for each user.

My initial impression was that you can get emails from many different users but the recipient was either one particular person or could be extracted from the text of the original email (again refer to example above).

**Comment by RafalGemex** [comment]
Both USER1 and USER2 are groups of contacts approx. 10 email addresses in each of the groups.

The Message Filter looks more like this

![Any of the following](https://user-images.githubusercontent.com/100281749/155315927-1e37dfb6-548e-48b7-97b6-805d1e3ee0f8.PNG)
IF FROM is USER1.1
IF FROM is USER1.2
IF FROM is USER1.3
IF FROM is USER1.4
IF FROM is USER1.5
IF FROM is USER1.6
IF FROM is USER1.7
IF FROM is USER1.8
IF FROM is USER1.9
IF FROM is USER1.10

THEN

FORWARD MESSAGE TO USER2.1; USER2.2; USER2.3; USER2.4; USER2.5; USER2.6; USER2.7; USER2.8; USER2.9; USER2.10

**Comment by RealRaven2000** [comment]
> Both USER1 and USER2 are groups of contacts approx. 10 email addresses in each of the groups.
> 
> The Message Filter looks more like this
> 
> ![Any of the following](https://user-images.githubusercontent.com/100281749/155315927-1e37dfb6-548e-48b7-97b6-805d1e3ee0f8.PNG) IF FROM is USER1.1 IF FROM is USER1.2 IF FROM is USER1.3 IF FROM is USER1.4 IF FROM is USER1.5 IF FROM is USER1.6 IF FROM is USER1.7 IF FROM is USER1.8 IF FROM is USER1.9 IF FROM is USER1.10
> 
> THEN
> 
> FORWARD MESSAGE TO USER2.1; USER2.2; USER2.3; USER2.4; USER2.5; USER2.6; USER2.7; USER2.8; USER2.9; USER2.10

ok, now if you have a _single_ filter like that you are forwarding the same email(s) to a group of different users (which are coded in the filter as an action as it looks right now).

Instead, you would do "forward using template x" and then template x would have a list of recipients hard coded as Cc (so that no user will see who else the email was sent to):

```
%header.delete(to)%
%header.set(cc,"USER1")%
%header.append(cc,"USER2")%
%header.append(cc,"USER3")%

```

Of course we cannot then add a salutation ("Dear USER1 / USER2" etc.) because processing of SmartTemplates happens when the email loads (not when it is sent off). Cc's are stripped at the server side, so there is no way to do this with an Add-on (extensions always run client side only).

(Edit: removed spaces from SmartTemplates commands!")

**Comment by RealRaven2000** [comment]
PS - when forwarding, it is also possible to use an address list to set the Cc headers since [issue 167](https://github.com/RealRaven2000/SmartTemplates/issues/167) was implemented. this way you do not have to edit the template to add / remove recipients - just add people to the Address Book list instead.

**Comment by RafalGemex** [comment]
I am still having problems setting this up.

Can you please kindly assist?

USER1: rafal@gemex.biz
USER2: rafaluklejewski@gemex.biz
Thunderbird: grarafka@gmail.com

And also I am not sure how to set it up with SmartTamplates

**Comment by RafalGemex** [comment]
Also, please note that there are two active filters.

First one:

[](https://user-images.githubusercontent.com/100281749/155315927-1e37dfb6-548e-48b7-97b6-805d1e3ee0f8.PNG)Any of the following
IF FROM is USER1.1
IF FROM is USER1.2
IF FROM is USER1.3
IF FROM is USER1.4
IF FROM is USER1.5
IF FROM is USER1.6
IF FROM is USER1.7
IF FROM is USER1.8
IF FROM is USER1.9
IF FROM is USER1.10

THEN

FORWARD MESSAGE TO USER2.1; USER2.2; USER2.3; USER2.4; USER2.5; USER2.6; USER2.7; USER2.8; USER2.9; USER2.10

The second one:
[](https://user-images.githubusercontent.com/100281749/155315927-1e37dfb6-548e-48b7-97b6-805d1e3ee0f8.PNG)Any of the following
IF FROM is USER2.1
IF FROM is USER2.2
IF FROM is USER2.3
IF FROM is USER2.4
IF FROM is USER2.5
IF FROM is USER2.6
IF FROM is USER2.7
IF FROM is USER2.8
IF FROM is USER2.9
IF FROM is USER2.10

THEN

FORWARD MESSAGE TO USER1.1; USER1.2; USER1.3; USER1.4; USER1.5; USER1.6; USER1.7; USER1.8; USER1.9; USER1.10

**Comment by RealRaven2000** [comment]
> I am still having problems setting this up.
> 
> Can you please kindly assist?
> 
> USER1: [rafal@gemex.biz](mailto:rafal@gemex.biz) USER2: [rafaluklejewski@gemex.biz](mailto:rafaluklejewski@gemex.biz) Thunderbird: [grarafka@gmail.com](mailto:grarafka@gmail.com)

You would set up the filter something like this:

![image](https://user-images.githubusercontent.com/6314674/155533729-e05b4d67-51b6-4868-ae97-d8acab45a673.png)

...but as we discussed it won't work with SmartTemplates (yet) - because SmartTemplates uses the Composer window.

Once we have the feature of integrating both Add-ons with each other it will look like this:

![image](https://user-images.githubusercontent.com/6314674/155534361-16497c51-2f50-4975-8f69-7a9b61cf2a98.png)

the filter would still contain the conditions From contains USER1.1, contains USER1.2 etc.
but the action would only be forward with smart template pointing to a HTML file.
The html file would then contain the command for filling the recipients with USER 2.1 USER 2.2 etc.

There for you will need 2 different templates. Important: the email will always be forwarded to all users in the to list, just like if you use the normal forwarding action (without template)

> 
> And also I am not sure how to set it up with SmartTemplates

Obviously you cannot set it up yet to **use SmartTemplates with a filter action** (this is the whole point of this issue). But you can set up a manual forwarding template for testing now. I will explain it over at the [SmartTemplates Issue](https://github.com/RealRaven2000/SmartTemplates/issues/173)

**Comment by RealRaven2000** [comment]
Built the user interface for selecting the template:

[filtaquilla-3.4pre45.zip](https://github.com/RealRaven2000/FiltaQuilla/files/8135091/filtaquilla-3.4pre45.zip)

to install, download the archive file and drag into Thunderbird Add-ons Manager.

**Comment by RealRaven2000** [comment]
To activate the new action (forwarding smart templates from a filter), open FiltaQuilla Options from Add-ons Manager (using the wrench symbol)

![image](https://user-images.githubusercontent.com/6314674/155590023-5e63fa83-70dd-4b35-b91d-58a8df948b5f.png)

**Comment by RafalGemex** [comment]
I can not select the template as it's looking for html file and the template is in .json format.

Can you please let me know how to set it up?

**Comment by RealRaven2000** [comment]
> I can not select the template as it's looking for html file and the template is in .json format.

Nope. **_External templates_** are not in json format, they never are. You're probably looking at a saved files of account templates in SmartTemplates, (this is a file that saves all settings [_Write / Reply / Forward_] for the currently selected mail account, including _templates_, _quote headers_ and options such as [x] _Use html_) which is not the same as an external template - the tooltip "Save Template.." may be confusing - note to self: relabel this better.

For the purpose of Implementing Forward from Filter you do not use the account template field you want an external html file:

![image](https://user-images.githubusercontent.com/6314674/155708853-0430dfc7-94ff-4a27-a720-812415b247bc.png)

Instead you need to use an external template (which is always HTML or maybe txt, but HTML is preferred). Once an external HTML template is used, the text in the template field for the account (above) will be completely ignored. Just cut the text there, insert into an empty text document (notepad or better notepad++) and save as HTML file. Let's call it "_forwardToList1.html_"

You can test the html file (and even attach to the forward menu permanently) using these steps:

- click the down chevron on the forward button, and select "configure menu items":
![image](https://user-images.githubusercontent.com/6314674/155709241-2d063bb3-8234-457e-85af-c2640e9cb287.png)

- this will open setting with the (external) list of forwarding templates (in the Account dropdown it has selected "templates" for you and in the tabs it has selected "forward"):
![image](https://user-images.githubusercontent.com/6314674/155709407-d13566a1-2af1-4a90-992a-9f62b0831980.png)

- click the red flashing button "Select template file from disk":
![image](https://user-images.githubusercontent.com/6314674/155709504-ef258d09-a6f3-446d-92aa-50e8aee216e5.png)

- open the "forwardToList1.html" file, then edit the title for the menu, and click the _Add_ button:
![image](https://user-images.githubusercontent.com/6314674/155709765-8204025d-46b1-4018-a515-c3f448b3dca6.png)

- now the new template shows up on the forward button:
![image](https://user-images.githubusercontent.com/6314674/155709835-e9de5c63-2de6-4be6-9cd0-cc31841ba373.png)

...this is how you can conveniently test your template and make everything perfect while I implement the automatic sending / reacting to the filter (WIP). And you can always edit the html and improve it, and test the new file by clicking the menu item "Forward To List 1" - no need to add it again. 

As soon as you use an external template (via one of the dropdown menus on the Write / Forward / Reply buttons) then the account template shown on the Account settings page will be ignored. However, the quote header may be used, if the "Quote Header" field is filled. So for the moment leave the quote header field empty. I might add a new command `%suppressQuoteHeaders%` later so that the template can completely ignore the quote header from the account while removing the original quote header ( --- on 13/2/22 Fred wrote: --- ) regardless of account settings.

Once the filter action is fully integrated, SmartTemplates will receive the template file name (HTML file!) and list of matching emails and carry out the forward / reply for you.



**Efforts So Far**

I have done about 2 days of work on this already, but I have a music production gig this weekend, so I am planning to continue it next week. If you want your company to officially fund some of this work, feel free to reach out via E-mail and we can work something out. 

In the end this feature could potentially largely benefit some other users, but my user numbers are very low (only 6,738 users) and monetization has been lackluster, so I really need to ramp up  marketing efforts to make it worth while.

**Comment by RealRaven2000** [comment]
Edit: in the previous comment I forgot to add to "click the Add button" in the 4th step, so the full workflow is this:

![image](https://user-images.githubusercontent.com/6314674/155713048-a2d5ad61-5db5-473c-97b7-cb7e3386070e.png)

1 - open file
2 - edit title
3 - click Add

You can also test any external HTML file without adding it to the dropdown permanently by selecting "Open SmartTemplates Template file..."
![image](https://user-images.githubusercontent.com/6314674/155713208-3d3ca25e-7590-4448-9a1d-339a43c37643.png)

for your purpose (repeated testing of the template until it works perfectly) having it on the menu is much better though.

**Comment by RealRaven2000** [comment]
Also added support for "Reply with SmartTemplate:"

[filtaquilla-3.4pre54.zip](https://github.com/RealRaven2000/FiltaQuilla/files/8156653/filtaquilla-3.4pre54.zip)

![image](https://user-images.githubusercontent.com/6314674/156056621-657eaae0-3b6e-4ffc-b8ab-395b0b4d9a48.png)

Since custom actions cannot be removed in Thunderbird, you may need to restart Thunderbird after installation to avoid duplication / outdated functions in the actions dropdown.

**Comment by RafalGemex** [comment]
Question: Does the template file should be an empty file with .html as an extension?

**Comment by RealRaven2000** [comment]
> Question: Does the template file should be an empty file with .html as an extension?

Yes, just use an empty file (technically it is just a html fragment, no need for DOCTYPE or any other HTML statments  unless you want to insert something like a `<p>` or `<h1>` etc.)
Just **add your SmartTemplate commands** such as the ones I described at:
https://github.com/RealRaven2000/SmartTemplates/issues/173#issuecomment-1054518575

For example:

```
%sig(none)%
%suppressQuoteHeaders%
%header.set(to,"axel.grude@gmail.com")%
%header.append(to,"realraven2000@hotmail.com")%
%header.append(to,"seawonkey@gmail.com")%
%header.delete(subject,"Fwd: ")%
```

This example will both suppress the signature and quote header, then set the recipients and remove "Fwd: " from the subject line.

**Comment by RafalGemex** [comment]
Also, Is there any chance to find and replace text in the body with the commands?

**Comment by RealRaven2000** [comment]
You can email me directly for these detailed questions that's probably going to be easier and faster.

**Comment by RealRaven2000** [comment]
> Also, Is there any chance to find and replace text in the body with the commands?

Also these questions are very specific to **SmartTemplates** so if you want to ask on the github please [ask them from here](https://github.com/RealRaven2000/SmartTemplates/issues/173).

**Comment by RafalGemex** [comment]
Can you find and replace text in the body of an email using the commands?

**Comment by RealRaven2000** [comment]
> Can you find and replace text in the body of an email using the commands?

Yes - I have replied on the SmartTemplates issue, but I also sent you an email directly with a full summary of everything you need to do, could you reply to that so I know you had a  chance to read this.

**Comment by RealRaven2000** [comment]
> Can you find and replace text in the body of an email using the commands?

did you get my direct email? You haven't replied so far...

**Comment by RealRaven2000** [comment]
Implemented in 3.4, released 18/03/2022
